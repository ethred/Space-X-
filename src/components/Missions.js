import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { toggleJoinStatus } from '../Redux/missions/missionSlice';
import './styles/Missions.css';

const Missions = () => {
  const { missions, loading, error } = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return (
      <div>
        Error:
        {' '}
        {error}
      </div>
    );
  }

  return (
    <table className="missions-table">
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <tr key={mission.mission_id}>
            <td>{mission.mission_name}</td>
            <td>{mission.description}</td>
            <td>
              <div className="status-column">
                <div className={`status-text ${mission.joined ? 'active-member' : 'not-member'}`}>
                  <p className="status-p">{mission.joined ? 'Active member' : 'Not a member'}</p>
                </div>
              </div>
            </td>
            <td>
              <button
                className={`status-button ${mission.joined ? 'leave' : ''}`}
                type="button"
                onClick={() => {
                  dispatch(toggleJoinStatus(mission.mission_id));
                }}
              >
                {mission.joined ? 'Leave Mission' : 'Join Mission'}
              </button>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default Missions;
