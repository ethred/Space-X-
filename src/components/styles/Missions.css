.missions-table {
  width: 100%;
  border-collapse: collapse;
  margin-bottom: 20px;
}

.missions-table th {
  border: 1px solid #ccc;
  padding: 10px;
  text-align: left;
  background-color: #f2f2f2;
}

.missions-table td {
  border: 1px solid #ccc;
  padding: 10px;
  text-align: left;
}

.missions-table th:nth-child(1),
.missions-table td:nth-child(1) {
  width: 150px;
}

.missions-table th:nth-child(2),
.missions-table td:nth-child(2) {
  width: 65vw;
}

.status-column {
  display: table;
  width: 100%;
}

.status-text {
  display: table-cell;
  padding: 2px;
  border-right: 1px solid #ccc;
  border-radius: 5px;
}

.status-button {
  border: 1px solid #ccc;
  background-color: white;
  font-weight: bold;
  cursor: pointer;
  color: black;
  padding: 5px 10px;
  transition: color 0.3s, border-color 0.3s, background-color 0.3s;
  border-radius: 5px;
}

.status-button.leave {
  border-color: red;
  color: red;
  background-color: white;
}

.active-member {
  background-color: turquoise;
}

.not-member {
  background-color: grey;
}

.status-p {
  margin: 0;
  text-align: center;
}
