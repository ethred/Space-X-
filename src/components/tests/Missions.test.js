import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Missions from '../Missions';

const mockStore = configureStore([]);
const store = mockStore({
  missions: {
    missions: [
      {
        mission_id: 1,
        mission_name: 'Mission 1',
        description: 'Description 1',
        joined: true,
      },
    ],
    loading: false,
    error: null,
  },
});

it('renders correctly', () => {
  const component = renderer.create(
    <Provider store={store}>
      <Missions />
    </Provider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
