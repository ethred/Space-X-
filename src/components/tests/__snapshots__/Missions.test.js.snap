// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="missions-table"
>
  <thead>
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Mission 1
      </td>
      <td>
        Description 1
      </td>
      <td>
        <div
          className="status-column"
        >
          <div
            className="status-text active-member"
          >
            <p
              className="status-p"
            >
              Active member
            </p>
          </div>
        </div>
      </td>
      <td>
        <button
          className="status-button leave"
          onClick={[Function]}
          type="button"
        >
          Leave Mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
